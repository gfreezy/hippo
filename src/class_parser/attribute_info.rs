mod predefined_attribute;
use nom::eof;
use nom::multi::length_data;
use nom::number::complete::{be_u16, be_u32};
use nom::IResult;

#[derive(Debug)]
pub struct AttributeInfo {
    pub attribute_name_index: u16,
    pub info: Vec<u8>,
}

pub fn parse_attribute_info(buf: &[u8]) -> IResult<&[u8], AttributeInfo> {
    let (left, attribute_name_index) = be_u16(buf)?;
    let (left, info) = length_data(be_u32)(left)?;
    Ok((
        left,
        AttributeInfo {
            attribute_name_index,
            info: info.to_vec(),
        },
    ))
}

#[cfg(test)]
mod tests {
    use crate::class_parser::attribute_info::parse_attribute_info;
    use insta::assert_debug_snapshot;
    use nom::multi::many_m_n;

    #[test]
    fn test_parse_attribute_info() {
        let buf = [0, 13, 0, 0, 0, 2, 0, 14];
        let parser = many_m_n(1, 1, parse_attribute_info);
        assert_debug_snapshot!(parser(&buf).unwrap());
    }
}
