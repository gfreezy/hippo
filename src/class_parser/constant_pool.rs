use cesu8::from_java_cesu8;
use enum_methods::EnumIsA;
use nom::multi::length_data;
use nom::number::complete::{be_f32, be_f64, be_i32, be_i64, be_u16, be_u8};
use nom::IResult;

const CONSTANT_CLASS: u8 = 7;
const CONSTANT_FIELDREF: u8 = 9;
const CONSTANT_METHODREF: u8 = 10;
const CONSTANT_INTERFACE_METHODREF: u8 = 11;
const CONSTANT_STRING: u8 = 8;
const CONSTANT_INTEGER: u8 = 3;
const CONSTANT_FLOAT: u8 = 4;
const CONSTANT_LONG: u8 = 5;
const CONSTANT_DOUBLE: u8 = 6;
const CONSTANT_NAME_AND_TYPE: u8 = 12;
const CONSTANT_UTF8: u8 = 1;
const CONSTANT_METHOD_HANDLE: u8 = 15;
const CONSTANT_METHOD_TYPE: u8 = 16;
const CONSTANT_INVOKE_DYNAMIC: u8 = 18;

#[derive(Debug, EnumIsA)]
pub(crate) enum ConstPoolInfo {
    ConstantClassInfo {
        name_index: u16,
    },
    ConstantFieldrefInfo {
        class_index: u16,
        name_and_type_index: u16,
    },
    ConstantStringInfo {
        string_index: u16,
    },
    ConstantIntegerInfo(i32),
    ConstantFloatInfo(f32),
    ConstantLongInfo(i64),
    ConstantDoubleInfo(f64),
    ConstantNameAndTypeInfo {
        name_index: u16,
        descriptor_index: u16,
    },
    ConstantUtf8Info(String),
    ConstantMethodHandleInfo {
        reference_kind: u8,
        reference_index: u16,
    },
    ConstantMethodTypeInfo {
        descriptor_index: u16,
    },
    ConstantInvokeDynamicInfo {
        bootstrap_method_attr_index: u16,
        name_and_type_index: u16,
    },
}

impl ConstPoolInfo {}

pub(crate) fn parse_const_pool_info(buf: &[u8]) -> IResult<&[u8], ConstPoolInfo> {
    use ConstPoolInfo::*;

    let (left, pool_tag) = be_u8(buf)?;
    match pool_tag {
        CONSTANT_CLASS => {
            let (left, name_index) = be_u16(left)?;
            Ok((left, ConstantClassInfo { name_index }))
        }
        CONSTANT_FIELDREF | CONSTANT_METHODREF | CONSTANT_INTERFACE_METHODREF => {
            let (left, class_index) = be_u16(left)?;
            let (left, name_and_type_index) = be_u16(left)?;
            Ok((
                left,
                ConstantFieldrefInfo {
                    class_index,
                    name_and_type_index,
                },
            ))
        }
        CONSTANT_STRING => {
            let (left, string_index) = be_u16(left)?;
            Ok((left, ConstantStringInfo { string_index }))
        }
        CONSTANT_INTEGER => {
            let (left, i) = be_i32(left)?;
            Ok((left, ConstantIntegerInfo(i)))
        }
        CONSTANT_FLOAT => {
            let (left, f) = be_f32(left)?;
            Ok((left, ConstantFloatInfo(f)))
        }
        CONSTANT_LONG => {
            let (left, i) = be_i64(left)?;
            Ok((left, ConstantLongInfo(i)))
        }
        CONSTANT_DOUBLE => {
            let (left, i) = be_f64(left)?;
            Ok((left, ConstantDoubleInfo(i)))
        }
        CONSTANT_NAME_AND_TYPE => {
            let (left, name_index) = be_u16(left)?;
            let (left, descriptor_index) = be_u16(left)?;
            Ok((
                left,
                ConstantNameAndTypeInfo {
                    name_index,
                    descriptor_index,
                },
            ))
        }
        CONSTANT_UTF8 => {
            let (left, bytes) = length_data(be_u16)(left)?;
            let s = from_java_cesu8(bytes).expect("invalid utf-8 string");
            Ok((left, ConstantUtf8Info(s.to_string())))
        }
        CONSTANT_METHOD_HANDLE => {
            let (left, reference_kind) = be_u8(left)?;
            let (left, reference_index) = be_u16(left)?;
            Ok((
                left,
                ConstantMethodHandleInfo {
                    reference_kind,
                    reference_index,
                },
            ))
        }
        CONSTANT_METHOD_TYPE => {
            let (left, descriptor_index) = be_u16(left)?;
            Ok((left, ConstantMethodTypeInfo { descriptor_index }))
        }
        CONSTANT_INVOKE_DYNAMIC => {
            let (left, bootstrap_method_attr_index) = be_u16(left)?;
            let (left, name_and_type_index) = be_u16(left)?;
            Ok((
                left,
                ConstantInvokeDynamicInfo {
                    bootstrap_method_attr_index,
                    name_and_type_index,
                },
            ))
        }
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use crate::class_parser::constant_pool::parse_const_pool_info;
    use insta::assert_debug_snapshot;
    use nom::multi::many_m_n;

    #[test]
    fn test_parse_const_pool_info() {
        let buf = [
            10, 0, 6, 0, 15, 9, 0, 16, 0, 17, 8, 0, 18, 10, 0, 19, 0, 20, 7, 0, 21, 7, 0, 22, 1, 0,
            6, 60, 105, 110, 105, 116, 62, 1, 0, 3, 40, 41, 86, 1, 0, 4, 67, 111, 100, 101, 1, 0,
            15, 76, 105, 110, 101, 78, 117, 109, 98, 101, 114, 84, 97, 98, 108, 101, 1, 0, 4, 109,
            97, 105, 110, 1, 0, 22, 40, 91, 76, 106, 97, 118, 97, 47, 108, 97, 110, 103, 47, 83,
            116, 114, 105, 110, 103, 59, 41, 86, 1, 0, 10, 83, 111, 117, 114, 99, 101, 70, 105,
            108, 101, 1, 0, 9, 77, 97, 105, 110, 46, 106, 97, 118, 97, 12, 0, 7, 0, 8, 7, 0, 23,
            12, 0, 24, 0, 25, 1, 0, 5, 72, 101, 108, 108, 111, 7, 0, 26, 12, 0, 27, 0, 28, 1, 0, 9,
            77, 97, 105, 110, 47, 77, 97, 105, 110, 1, 0, 16, 106, 97, 118, 97, 47, 108, 97, 110,
            103, 47, 79, 98, 106, 101, 99, 116, 1, 0, 16, 106, 97, 118, 97, 47, 108, 97, 110, 103,
            47, 83, 121, 115, 116, 101, 109, 1, 0, 3, 111, 117, 116, 1, 0, 21, 76, 106, 97, 118,
            97, 47, 105, 111, 47, 80, 114, 105, 110, 116, 83, 116, 114, 101, 97, 109, 59, 1, 0, 19,
            106, 97, 118, 97, 47, 105, 111, 47, 80, 114, 105, 110, 116, 83, 116, 114, 101, 97, 109,
            1, 0, 7, 112, 114, 105, 110, 116, 108, 110, 1, 0, 21, 40, 76, 106, 97, 118, 97, 47,
            108, 97, 110, 103, 47, 83, 116, 114, 105, 110, 103, 59, 41, 86,
        ];
        let parser = many_m_n(28, 28, parse_const_pool_info);
        assert_debug_snapshot!(parser(&buf).expect("parse const pool"));
    }
}
