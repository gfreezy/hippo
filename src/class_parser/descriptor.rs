use nom::branch::alt;
use nom::bytes::complete::tag;
use nom::character::complete::none_of;
use nom::multi::{many0, many1};
use nom::IResult;

fn reference_type(buf: &str) -> IResult<&str, String> {
    let (buf, _) = tag("L")(buf)?;
    let (buf, class_name) = many1(none_of(";["))(buf)?;
    let (buf, _) = tag(";")(buf)?;
    Ok((
        buf,
        format!("L{};", class_name.into_iter().collect::<String>()),
    ))
}

fn base_type(buf: &str) -> IResult<&str, String> {
    let (buf, d) = alt((
        tag("B"),
        tag("C"),
        tag("D"),
        tag("F"),
        tag("I"),
        tag("J"),
        tag("S"),
        tag("Z"),
    ))(buf)?;
    Ok((buf, d.to_string()))
}

fn array_type(buf: &str) -> IResult<&str, String> {
    let (buf, array) = many1(tag("["))(buf)?;
    let (buf, ty) = alt((base_type, reference_type))(buf)?;
    Ok((buf, format!("{}{}", array.join(""), ty)))
}

pub fn field_descriptor(buf: &str) -> IResult<&str, String> {
    alt((base_type, reference_type, array_type))(buf)
}

fn void_descriptor(buf: &str) -> IResult<&str, String> {
    let (buf, v) = tag("V")(buf)?;
    Ok((buf, v.to_string()))
}

pub fn method_descriptor(buf: &str) -> IResult<&str, (Vec<String>, String)> {
    let (buf, _) = tag("(")(buf)?;
    let (buf, parameters) = many0(field_descriptor)(buf)?;
    let (buf, _) = tag(")")(buf)?;
    let (buf, return_descriptor) = alt((void_descriptor, field_descriptor))(buf)?;
    Ok((buf, (parameters, return_descriptor)))
}

#[cfg(test)]
mod tests {
    use super::{field_descriptor, method_descriptor};

    #[test]
    fn test_descriptor() {
        let (_, field) = field_descriptor("I").unwrap();
        assert_eq!(field, "I");
        let (_, field) = field_descriptor("Ljava/lang/Thread;").unwrap();
        assert_eq!(field, "java/lang/Thread");
        let (_, method) = method_descriptor("(IDLjava/lang/Thread;)Ljava/lang/Object;").unwrap();
        assert_eq!(
            method,
            (
                vec![
                    "I".to_string(),
                    "D".to_string(),
                    "java/lang/Thread".to_string()
                ],
                "java/lang/Object".to_string()
            )
        );
        let (_, method) =
            method_descriptor("(IDLjava/lang/Thread;[[[Ljava/lang/Thread;)Ljava/lang/Object;")
                .unwrap();
        assert_eq!(
            method,
            (
                vec![
                    "I".to_string(),
                    "D".to_string(),
                    "java/lang/Thread".to_string(),
                    "[[[java/lang/Thread".to_string(),
                ],
                "java/lang/Object".to_string()
            )
        )
    }
}
