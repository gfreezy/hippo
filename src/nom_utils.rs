use nom::error::ParseError;
use nom::multi::many_m_n;
use nom::{IResult, InputLength, InputTake, ToUsize};

pub fn length_many<I, O, N, E, F, G>(f: F, g: G) -> impl Fn(I) -> IResult<I, Vec<O>, E>
where
    I: Clone + InputLength + InputTake + PartialEq,
    N: Copy + ToUsize,
    F: Fn(I) -> IResult<I, N, E>,
    G: Fn(I) -> IResult<I, O, E> + Copy,
    E: ParseError<I>,
{
    move |buf| {
        let (buf, num) = f(buf)?;
        many_m_n(num.to_usize(), num.to_usize(), g)(buf)
    }
}
