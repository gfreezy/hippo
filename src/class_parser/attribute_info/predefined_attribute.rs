use crate::class_parser::attribute_info::AttributeInfo;

pub struct ConstantValueAttribute {
    constant_value_index: u16,
}

pub struct ExceptionHandler {
    start_pc: u16,
    end_pc: u16,
    handler_pc: u16,
    catch_type: u16,
}

pub struct CodeAttribute {
    max_stack: u16,
    max_locals: u16,
    code: Vec<u8>,
    exception_table: Vec<ExceptionHandler>,
    attributes: Vec<AttributeInfo>,
}
