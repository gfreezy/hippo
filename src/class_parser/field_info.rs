use crate::class_parser::attribute_info::{parse_attribute_info, AttributeInfo};
use nom::multi::{length_data, many_m_n};
use nom::number::complete::{be_f32, be_f64, be_i32, be_i64, be_u16, be_u32, be_u8};
use nom::IResult;

#[derive(Debug)]
pub struct FieldInfo {
    access_flags: u16,
    name_index: u16,
    descriptor_index: u16,
    attributes: Vec<AttributeInfo>,
}

pub fn parse_field_info(buf: &[u8]) -> IResult<&[u8], FieldInfo> {
    let (left, access_flags) = be_u16(buf)?;
    let (left, name_index) = be_u16(left)?;
    let (left, descriptor_index) = be_u16(left)?;
    let (left, attributes_count) = be_u16(left)?;
    let (left, attributes) = many_m_n(
        attributes_count as usize,
        attributes_count as usize,
        parse_attribute_info,
    )(left)?;
    Ok((
        left,
        FieldInfo {
            access_flags,
            name_index,
            descriptor_index,
            attributes,
        },
    ))
}
