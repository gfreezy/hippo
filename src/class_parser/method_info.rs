use crate::class_parser::attribute_info::{parse_attribute_info, AttributeInfo};
use crate::class_parser::constant_pool::ConstPoolInfo;
use crate::nom_utils::length_many;
use nom::number::complete::be_u16;
use nom::IResult;

#[derive(Debug)]
pub struct MethodInfo {
    access_flags: u16,
    name_index: u16,
    descriptor_index: u16,
    attributes: Vec<AttributeInfo>,
}

pub fn parse_method_info<'a>(
    const_pools: &Vec<ConstPoolInfo>,
    buf: &'a [u8],
) -> IResult<&'a [u8], MethodInfo> {
    let (left, access_flags) = be_u16(buf)?;
    let (left, name_index) = be_u16(left)?;
    let (left, descriptor_index) = be_u16(left)?;
    let (left, attributes) =
        length_many(be_u16, |buf| parse_attribute_info(const_pools, buf))(left)?;
    Ok((
        left,
        MethodInfo {
            access_flags,
            name_index,
            descriptor_index,
            attributes,
        },
    ))
}
