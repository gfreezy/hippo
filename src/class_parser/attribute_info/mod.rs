#[macro_use]
pub mod predefined_attribute;

use crate::class_parser::attribute_info::predefined_attribute::{
    parse_predefined_attribute, PredefinedAttribute,
};
use crate::class_parser::constant_pool::ConstPool;
use nom::number::complete::{be_u16, be_u32};
use nom::IResult;

#[derive(Debug)]
pub struct AttributeInfo {
    pub attribute_name_index: u16,
    pub attribute: PredefinedAttribute,
}

pub fn parse_attribute_info<'a>(
    const_pool: &ConstPool,
    buf: &'a [u8],
) -> IResult<&'a [u8], AttributeInfo> {
    let (buf, attribute_name_index) = be_u16(buf)?;
    let (buf, _length) = be_u32(buf)?;
    let attr_name = const_pool.get_utf8_string_at(attribute_name_index);
    let (buf, attr) = parse_predefined_attribute(attr_name, const_pool, buf)?;
    Ok((
        buf,
        AttributeInfo {
            attribute_name_index,
            attribute: attr,
        },
    ))
}
