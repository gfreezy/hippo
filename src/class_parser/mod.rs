pub mod attribute_info;
pub mod constant_pool;
pub mod field_info;
pub mod method_info;

use crate::class_parser::attribute_info::{parse_attribute_info, AttributeInfo};
use crate::class_parser::constant_pool::{parse_const_pool_info, ConstPool};
use crate::class_parser::field_info::{parse_field_info, FieldInfo};
use crate::class_parser::method_info::{parse_method_info, MethodInfo};
use crate::nom_utils::length_many;
use anyhow::{ensure, Result};
use nom::bytes::complete::tag;
use nom::eof;
use nom::multi::many_m_n;
use nom::number::complete::be_u16;
use nom::IResult;

const MAGIC_NUMBER: u32 = 0xCAFEBABE;

const ACC_PUBLIC: u16 = 0x0001;
const ACC_PRIVATE: u16 = 0x0002;
const ACC_PROTECTED: u16 = 0x0004;
const ACC_STATIC: u16 = 0x0008;
const ACC_FINAL: u16 = 0x0010;
const ACC_SYNCHRONIZED: u16 = 0x0020;
const ACC_SUPER: u16 = 0x0020;
const ACC_BRIDGE: u16 = 0x0040;
const ACC_VOLATILE: u16 = 0x0040;
const ACC_VARARGS: u16 = 0x0080;
const ACC_TRANSIENT: u16 = 0x0080;
const ACC_NATIVE: u16 = 0x0100;
const ACC_INTERFACE: u16 = 0x0200;
const ACC_ABSTRACT: u16 = 0x0400;
const ACC_STRICT: u16 = 0x0800;
const ACC_SYNTHETIC: u16 = 0x1000;
const ACC_ANNOTATION: u16 = 0x2000;
const ACC_ENUM: u16 = 0x4000;

#[derive(Debug)]
struct ClassFile {
    minor_version: u16,
    major_version: u16,
    constant_pool: ConstPool,
    access_flags: u16,
    this_class: u16,
    super_class: u16,
    interfaces: Vec<u16>,
    fields: Vec<FieldInfo>,
    methods: Vec<MethodInfo>,
    attributes: Vec<AttributeInfo>,
}

fn is_bit_set(num: u16, flag: u16) -> bool {
    num & flag != 0
}

fn is_bit_clear(num: u16, flag: u16) -> bool {
    num & flag == 0
}

impl ClassFile {
    pub fn new(
        minor_version: u16,
        major_version: u16,
        constant_pool: ConstPool,
        access_flags: u16,
        this_class: u16,
        super_class: u16,
        interfaces: Vec<u16>,
        fields: Vec<FieldInfo>,
        methods: Vec<MethodInfo>,
        attributes: Vec<AttributeInfo>,
    ) -> Result<ClassFile> {
        let class_file = ClassFile {
            minor_version,
            major_version,
            constant_pool,
            access_flags,
            this_class,
            super_class,
            interfaces,
            fields,
            methods,
            attributes,
        };
        class_file.validate_access_flags()?;
        class_file.validate_this_class()?;
        //        class_file.validate_super_class()?;
        class_file.validate_interfaces()?;

        Ok(class_file)
    }

    fn validate_access_flags(&self) -> Result<()> {
        let access_flags = self.access_flags;
        if is_bit_set(access_flags, ACC_INTERFACE) {
            ensure!(
                is_bit_set(access_flags, ACC_ABSTRACT),
                "ACC_ABSTRACT is set"
            );
            ensure!(
                is_bit_clear(access_flags, ACC_FINAL)
                    && is_bit_clear(access_flags, ACC_SUPER)
                    && is_bit_clear(access_flags, ACC_ENUM),
                "access flags"
            );
        } else {
            ensure!(is_bit_clear(access_flags, ACC_ANNOTATION), "access flags");
            ensure!(
                !(is_bit_set(access_flags, ACC_FINAL) && is_bit_set(access_flags, ACC_ABSTRACT)),
                "access flags"
            );
        }
        Ok(())
    }

    fn validate_this_class(&self) -> Result<()> {
        ensure!(
            self.constant_pool.is_class_at(self.this_class),
            "validate this class"
        );
        Ok(())
    }

    fn validate_super_class(&self) -> Result<()> {
        if self.super_class == 0 {
            return Ok(());
        }
        ensure!(
            self.constant_pool.is_class_at(self.this_class),
            "validate super class"
        );
        Ok(())
    }

    fn validate_interfaces(&self) -> Result<()> {
        for interface in &self.interfaces {
            ensure!(
                self.constant_pool.is_class_at(*interface),
                "interface is not class"
            );
        }
        Ok(())
    }
}

fn parse_class_file(buf: &[u8]) -> IResult<&[u8], ClassFile> {
    let (left, _) = tag(&MAGIC_NUMBER.to_be_bytes()[..])(buf)?;
    let (left, minor_version) = be_u16(left)?;
    let (left, major_version) = be_u16(left)?;
    let (left, constant_pool_count) = be_u16(left)?;
    let (left, constant_pool_infos) =
        parse_items(constant_pool_count, parse_const_pool_info, left)?;
    let constant_pool = ConstPool::new(constant_pool_infos);
    let (left, access_flags) = be_u16(left)?;
    let (left, this_class) = be_u16(left)?;
    let (left, super_class) = be_u16(left)?;
    let (left, interfaces) = length_many(be_u16, be_u16)(left)?;
    let (left, fields) = length_many(be_u16, |buf| parse_field_info(&constant_pool, buf))(left)?;
    let (left, methods) = length_many(be_u16, |buf| parse_method_info(&constant_pool, buf))(left)?;
    let (left, attributes) =
        length_many(be_u16, |buf| parse_attribute_info(&constant_pool, buf))(left)?;
    let (left, _) = eof!(left,)?;

    Ok((
        left,
        ClassFile::new(
            minor_version,
            major_version,
            constant_pool,
            access_flags,
            this_class,
            super_class,
            interfaces,
            fields,
            methods,
            attributes,
        )
        .expect("parse class file"),
    ))
}

fn parse_items<I, O, F>(size: u16, parser: F, buf: I) -> IResult<I, Vec<O>>
where
    I: Clone + PartialEq,
    F: Fn(I) -> IResult<I, O>,
{
    if size > 0 {
        many_m_n(size as usize - 1, size as usize - 1, parser)(buf)
    } else {
        Ok((buf, Vec::new()))
    }
}

#[cfg(test)]
mod tests {
    use crate::class_parser::parse_class_file;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_class_file() -> anyhow::Result<()> {
        let data = [
            202, 254, 186, 190, 0, 0, 0, 52, 0, 29, 10, 0, 6, 0, 15, 9, 0, 16, 0, 17, 8, 0, 18, 10,
            0, 19, 0, 20, 7, 0, 21, 7, 0, 22, 1, 0, 6, 60, 105, 110, 105, 116, 62, 1, 0, 3, 40, 41,
            86, 1, 0, 4, 67, 111, 100, 101, 1, 0, 15, 76, 105, 110, 101, 78, 117, 109, 98, 101,
            114, 84, 97, 98, 108, 101, 1, 0, 4, 109, 97, 105, 110, 1, 0, 22, 40, 91, 76, 106, 97,
            118, 97, 47, 108, 97, 110, 103, 47, 83, 116, 114, 105, 110, 103, 59, 41, 86, 1, 0, 10,
            83, 111, 117, 114, 99, 101, 70, 105, 108, 101, 1, 0, 9, 77, 97, 105, 110, 46, 106, 97,
            118, 97, 12, 0, 7, 0, 8, 7, 0, 23, 12, 0, 24, 0, 25, 1, 0, 5, 72, 101, 108, 108, 111,
            7, 0, 26, 12, 0, 27, 0, 28, 1, 0, 9, 77, 97, 105, 110, 47, 77, 97, 105, 110, 1, 0, 16,
            106, 97, 118, 97, 47, 108, 97, 110, 103, 47, 79, 98, 106, 101, 99, 116, 1, 0, 16, 106,
            97, 118, 97, 47, 108, 97, 110, 103, 47, 83, 121, 115, 116, 101, 109, 1, 0, 3, 111, 117,
            116, 1, 0, 21, 76, 106, 97, 118, 97, 47, 105, 111, 47, 80, 114, 105, 110, 116, 83, 116,
            114, 101, 97, 109, 59, 1, 0, 19, 106, 97, 118, 97, 47, 105, 111, 47, 80, 114, 105, 110,
            116, 83, 116, 114, 101, 97, 109, 1, 0, 7, 112, 114, 105, 110, 116, 108, 110, 1, 0, 21,
            40, 76, 106, 97, 118, 97, 47, 108, 97, 110, 103, 47, 83, 116, 114, 105, 110, 103, 59,
            41, 86, 0, 33, 0, 5, 0, 6, 0, 0, 0, 0, 0, 2, 0, 1, 0, 7, 0, 8, 0, 1, 0, 9, 0, 0, 0, 29,
            0, 1, 0, 1, 0, 0, 0, 5, 42, 183, 0, 1, 177, 0, 0, 0, 1, 0, 10, 0, 0, 0, 6, 0, 1, 0, 0,
            0, 3, 0, 9, 0, 11, 0, 12, 0, 1, 0, 9, 0, 0, 0, 37, 0, 2, 0, 1, 0, 0, 0, 9, 178, 0, 2,
            18, 3, 182, 0, 4, 177, 0, 0, 0, 1, 0, 10, 0, 0, 0, 10, 0, 2, 0, 0, 0, 5, 0, 8, 0, 6, 0,
            1, 0, 13, 0, 0, 0, 2, 0, 14,
        ];

        let (buf, class) = parse_class_file(&data).expect("parse class");
        assert_debug_snapshot!((buf, class));
        Ok(())
    }
}
