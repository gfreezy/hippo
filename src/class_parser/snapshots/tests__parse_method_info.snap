---
source: src/class_parser/method_info.rs
expression: parser(&buf).unwrap()
---
(
    [],
    [
        MethodInfo {
            access_flags: 1,
            name_index: 7,
            descriptor_index: 8,
            attributes: [
                AttributeInfo {
                    attribute_name_index: 9,
                    info: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        0,
                        0,
                        5,
                        42,
                        183,
                        0,
                        1,
                        177,
                        0,
                        0,
                        0,
                        1,
                        0,
                        10,
                        0,
                        0,
                        0,
                        6,
                        0,
                        1,
                        0,
                        0,
                        0,
                        3,
                    ],
                },
            ],
        },
        MethodInfo {
            access_flags: 9,
            name_index: 11,
            descriptor_index: 12,
            attributes: [
                AttributeInfo {
                    attribute_name_index: 9,
                    info: [
                        0,
                        2,
                        0,
                        1,
                        0,
                        0,
                        0,
                        9,
                        178,
                        0,
                        2,
                        18,
                        3,
                        182,
                        0,
                        4,
                        177,
                        0,
                        0,
                        0,
                        1,
                        0,
                        10,
                        0,
                        0,
                        0,
                        10,
                        0,
                        2,
                        0,
                        0,
                        0,
                        5,
                        0,
                        8,
                        0,
                        6,
                    ],
                },
            ],
        },
    ],
)
